[1단계 - 출석 구현] 체체(김진영) 미션 제출합니다.

CheChe903 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/3947)에 셀프 체크를 작성했나요


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
메서드 명이 불분명한 게 많습니다.
객체의 책임이 올바른지 확실하지 않습니다.
인덴트가 2가 넘어가는 것들이 있습니다.
메서드 길이가 10라인이 넘는 것이 있습니다.

<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

1. DTO 사용에 대하여 궁금합니다. 제가 생각할 때 해당 두 DTO가 들어가야 좀 더 복잡하지 않는 구조를 갖는다고 생각하였습니다.
리뷰어님의 생각이 궁금합니다!



<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

rladmstn committed:
docs: README 작성

rladmstn committed:
docs: README 작성

rladmstn committed:
docs: 출석 csv 추가

rladmstn committed:
feat: Main 클래스 추가

rladmstn committed:
feat(FileReader): 파일 읽어오는 유틸 클래스 구현

rladmstn committed:
feat(AttendanceChecker): 출석 상태 지정 로직 구현

rladmstn committed:
feat(Attendance): 출석부 구현

출석부를 만들어, 해당 출석부에 없는 이름이 들어온다면 예외를 발생합니다.

rladmstn committed:
feat: 인프라 구축

rladmstn committed:
refactor(AttendanceChecker): 유틸 클래스로 변경

rladmstn committed:
refactor: 도메인 이름 변경

Attendance -> AttendanceBook
AttendanceRecord -> AttendanceRepository
Record -> Attendance

rladmstn committed:
feat(AttendanceBook): 출석부 생성 구현

rladmstn committed:
refactor(AttendanceReader): 출석 초기화 로직 수정

기존에는 파일 리더를 통해 출석 기록만 가져오던 로직을 DTO를 활용하여 출석부와 출결 기록을 함께 가져옵니다.
그에따른 파일들도 수정합니다.

rladmstn committed:
refactor: 컨벤션 진행

rladmstn committed:
feat(Attendance): 출결 기록 비교 로직 구현

비교 조건
1. 닉네임
2. 출석 날짜
같은 것이 있다면 true를 반환하고, 없다면 false를 반환합니다.

rladmstn committed:
feat(AttendanceRepository): 출결 기록 추가 로직 구현

rladmstn committed:
feat: 출석 확인 기능 구현

rladmstn committed:
docs: README 최신화

rladmstn committed:
feat(AttendanceChecker): 주말 출석 검증 로직 구현

rladmstn committed:
feat(AttendanceChecker): 출석 가능 시간 검증 구현

rladmstn committed:
feat(AttendanceRepository): 출석 수정 기능 구현

rladmstn committed:
feat(AttendanceController): 출석 수정 기능 구현

컨트롤러에서 제어하도록 합니다.

rladmstn committed:
refactor(Attendance): 날짜 검증 및 출결 상태 반환 로직 책임 이동

AttendanceChecker의 책임을 Attendance로 이동합니다.

rladmstn committed:
refactor(Time): 시간 객체 일급 컬렉션 적용

3,4번 기능을 위해 LocalDate를 가진 Time 객체를 생성합니다.
그에 따른 수정사항들도 커밋합니다.

rladmstn committed:
fix(Attendance): 기능 요구사항 반영

기존에는 0~5분 지각, 6분부터 결석으로 처리하였습니다.
이제는 ~5분까지 출석, ~30분까지 지각으로 처리합니다.

rladmstn committed:
feat: 3번 기능 정렬 추가 및 존재하지 않는 출석 초기화

1. 3번 기능을 위해, 존재하지 않는 출석을 List에 추가합니다.
2. 3번 기능에 대해서 출력하기 전 정렬 후 리턴합니다.

rladmstn committed:
test(AttendanceRepository): 테스트 수정

rladmstn committed:
fix(OutputView): 출석 수정 결과 출력 오류 수정

rladmstn committed:
feat(AttendanceRepository): 크루 네임에 따른 크루 학적 상태 구하는 로직 구현

출석, 결석, 지각의 수를 구하고 그에 따른 학적 상태를 반환하는 로직을 구현합니다.

rladmstn committed:
feat: 3번 기능 컨트롤러에 추가

rladmstn committed:
feat: 4번 기능 구현

제적 위험자들을 출력하는 기능을 구현합니다.

rladmstn committed:
feat: 재입력 구현 및 버그 수정

1. 시간 수정 안되던 버그 수정합니다.
2. 기능을 입력받을 때 int로 받던 것을 String으로 받습니다.
3. 개행을 추가합니다.
4. 시간 객체를 생성할 때 올바른 시각과 분을 입력하도록 합니다.

rladmstn committed:
refactor(AttendanceController): 입력 메서드 분리

rladmstn committed:
refactor: Time 레코드로 변경 및 Parser 유틸 클래스 생성

1. Time 클래스를 레코드로 변경합니다.
2. Parser 유틸 클래스를 생성합니다.

rladmstn committed:
refactor(AttendanceRepository): 메서드 리팩토링

rladmstn committed:
refactor: 출결 상태 및 학적 상태 ENUM 추가

출결 상태 및 학적 상태를 ENUM으로 바꾸면서
해당 값들을 계산하는 로직을 해당 ENUM들로 리팩터링합니다.

rladmstn committed:
refactor: 파라미터 final 추가

rladmstn committed:
docs: README 진행 사항 업데이트 및 공통 요구사항 수정

rladmstn committed:
fix(AttendanceStatus): 오타 수정

rladmstn committed:
refactor(Time): 등교 가능 시간 검증 메서드 리팩터링

등교 가능 날짜 및 시간 검증을 한 번에 하던 validatePossibleTime 메서드를
등교 가능 날짜 검증 메서드와 등교 가능 시간 검증 메서드로 분리하였습니다.

hyeonic commented on README.md:
```diff
@@ -1,3 +1,44 @@
+### 공통
+
+- [x] 잘못된 입력이 들어올 경우, 예외 메세지를 출력한 후 기능을 재입력 받는다.
+
+### 출석 확인
+
+- [x] 올바른 크루원의 닉네임을 입력한다.
+- [x] 24시간 형식으로 등교 시간을 입력한다.
+    - [x] 교육 시간은 월요일은 13:00 - 18:00, 화요일~금요일은 10:00 - 18:00이다.
+    - [x] 주말에는 출석을 받지 않는다.
+    - [x] 공휴일에는 출석을 받지 않는다.
+    - [x] 캠퍼스 운영 시간 (08:00 ~ 23:00) 외 출석은 인정되지 않는다.
+- [x] 출석이 인정되면 출석 기록을 출력한다.
+    - [x] 해당 요일의 시작 시각으로부터 5분 초과는 지각으로 간주한다.
+    - [x] 해당 요일의 시작 시각으로부터 30분 초과는 결석으로 간주한다.
+    - [x] 등교하지 않아 출석 기록이 없는 날은 결석으로 간주한다.
+- [x] 이미 출석한 경우, 다시 출석할 수 없으며 수정 기능을 이용하도록 안내한다.
+
+### 출석 수정
+
+- [x] 올바른 크루원의 닉네임을 입력한다.
+- [x] 수정 가능한 출석 날짜를 입력한다.
+- [x] 24시간 형식으로 등교 시간을 입력한다.
+- [x] 수정 후에는 변경 전과 변경 후의 출석 기록을 출력한다.
+- [x] 출석 기록을 수정할 수 있다.
+    - [x] 출석 기록이 없는 날에 수정하지 못한다.
+
+### 크루 별 출석 기록 확인
+
+- [x] 올바른 크루원의 닉네임을 입략한다.
+- [x] 전날까지의 해당 크루의 출석 기록을 출력한다.
+
+### 제적 위험자 확인
+
+- [x] 제적 위험자를 파악한다.
+    - [x] 경고 대상자: 결석 2회 이상
+    - [x] 면담 대상자: 결석 3회 이상
+    - [x] 제적 대상자: 결석 5회 초과
+- [x] 제적 대상자, 면담 대상자, 경고 대상자 순으로 출력한다.
+- [x] 지각을 결석으로 간주하여 내림차순한다. 같다면 닉네임으로 오름차순한다.
```
요구사항을 잘 정의해주셨네요 👍

CheChe903 replied:
감사합니다 ~!

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+            }
+
+        } while (true);
+    }
```
depth 1로 줄여볼까요??

CheChe903 replied:
매트, 이 부분에 스스로 좀 많이 생각해보았는데, 이러한 경우에 try-catch는 없앨 수 있지만 while, if는 분리하지 못했어요.
1로 줄일 수 있는 방법에 대한 힌트를 주면 더 생각해볼게요!

hyeonic replied:
콘솔 특성 상 입출력 부분은 depth 줄이기가 특히 어려울 수 있을 것 같아서 이정도면 충분한 것 같아요 👍

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+                throw new IllegalArgumentException("[ERROR] 올바른 기능을 입력해주세요.");
+            } catch (IllegalArgumentException e) {
+                outputView.printErrorMessage(e.getMessage());
+            }
```
정상 케이스에서도 에러가 발생하는 것 같아요!

![image](https://github.com/user-attachments/assets/e8fca35d-805b-46cb-8b61-4284eaab43ae)

CheChe903 replied:
체체라는 것은 main에서 가동시 없는 크루 이름입니다!! 
출석부를 만드는 것은 attendances.csv파일에 존재하는 크루를 기준으로 출석부를 만들기에,
해당 파일에 체체라는 크루는 없어 올바른 동작입니다!
다만, input이 두 번 되는 오류를 고쳤습니다!

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        return functionValue.equals("Q");
+    }
```
어떻게 하면 반복된 if문을 제거하면서 10줄 이내로 줄여볼 수 있을까요?? 조건과 해당 조건에 수행할 행위를 추상화할 순 없을까요??

CheChe903 replied:
행위에 대한 추상화를 진행해보았어요.

`
 private Map<String, AttendanceFunction> initializeFunctions() {
        return Map.of(
                CHECK_FUNCTION, new AttendanceCheckFunction(attendanceController),
                MODIFY_FUNCTION, new AttendanceModifyFunction(attendanceController),
                HISTORY_FUNCTION, new AttendanceHistoryFunction(attendanceController),
                RISK_OF_EXPULSION_FUNCTION, new RiskOfExpulsionFunction(attendanceController),
                QUIT_FUNCTION, new QuitFunction()
        );
    }
`

    
텍스트가 이상하게 보이긴 하지만,
현재 저는 각각 기능을 클래스로 만들어 execute를 진행했습니다. 매트가 생각한 방법이 이 방법이 맞는 지 궁금합니다!
또한, 패키지를 domain에 두는 것이 나은지, 아니면 다른 추천해주실만 한 패키지들이 있을까요 ?(Function에 대해)

hyeonic replied:
https://github.com/woowacourse/java-attendance/pull/38#discussion_r1966654406에 답변 남겼습니다!

hyeonic replied:
추가로 markdown 문법 "```" 활용하면 코드도 읽기 쉽게 전달할 수 있을 것 같아요!

hyeonic commented on src/test/java/attendance/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,85 @@
+                .hasMessage("[ERROR] 출석부에 없는 크루원입니다.");
+    }
+
+    private static Attendance makeAbsentAttendance(String name, int year, int month, int day) {
+        return new Attendance(name, new Time(LocalDate.of(year, month, day), "18", "00", true));
+    }
```
메서드 배치를 적절히 시키는 것도 중요한 코딩 컨벤션이라 생각해요! [클린 코드 5장 형식 맞추기](https://amazingguni.github.io/blog/2016/05/Clean-Code-5-%ED%98%95%EC%8B%9D-%EB%A7%9E%EC%B6%94%EA%B8%B0)를 참고해보거나 체체가 세운 기준으로 일관성을 맞춰보는 건 어떨까요??

CheChe903 replied:
시간이 없어 세세한 부분이 부족했던 것 같습니다 ㅠ.ㅠ

hyeonic commented on src/test/java/attendance/domain/AttendanceTest.java:
```diff
@@ -0,0 +1,85 @@
+
+        attendance1 = new Attendance("체체", new Time(localDate1, hour, minute, false));
+        attendance2 = new Attendance("체체", new Time(localDate2, hour, minute, false));
+
+
```
불필요한 공백은 제거되어도 좋을 것 같아요.

hyeonic commented on src/main/java/attendance/utils/FileReader.java:
```diff
@@ -0,0 +1,26 @@
+    private FileReader() {
+    }
+
+    public static List<String> parseToFile(final String filePath) {
+        try (BufferedReader reader = new BufferedReader(new java.io.FileReader(filePath))) {
+
+            List<String> readingFileResult = new ArrayList<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                readingFileResult.add(line);
+            }
+
+            return readingFileResult;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
try-with-resource를 적절히 활용하였네요 👍 어떤 장점이 있을까요??

CheChe903 replied:
close를 해주지 않아도 누수되지 않는다는 장점이 있습니다 :)

hyeonic commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,47 @@
+        return crewName;
+    }
+
+    public String getAttendanceStatus() {
+        if (attendanceTime.getDayOfWeek().equals("월요일")) { // 월요일
+
+            return AttendanceStatus.checkStatusWithCondition(attendanceTime, 13, 5, 30);
+        }
+
+        return AttendanceStatus.checkStatusWithCondition(attendanceTime, 10, 5, 30);
+    }
```
월요일 출석과 해당하는 기준 시간이 매우 중요해보이는데요, 적절히 상수 추출 하는 방향은 어떠신가요??

CheChe903 replied:
해당 부분도 시간이 부족하여 진행하지 못했습니다 !!

hyeonic commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public List<CrewNameAndAcademicStatusDTO> getCrewAtRiskOfExpulsion(final AttendanceRepository attendanceRepository,
+                                                                       final String academicStatus) {
+
+        return names.stream()
+                .map(attendanceRepository::getAcademicStatusByName)
+                .filter(dto -> dto.academicStatus().equals(academicStatus))
+                .sorted(Comparator.comparing(CrewNameAndAcademicStatusDTO::crewName))
+                .toList();
+
+    }
```
> DTO 사용에 대하여 궁금합니다. 제가 생각할 때 해당 두 DTO가 들어가야 좀 더 복잡하지 않는 구조를 갖는다고 생각하였습니다.
리뷰어님의 생각이 궁금합니다!

먼저 체체가 생각하는 DTO의 정의는 무엇인가요??

CheChe903 replied:
제가 생각하는 DTO는 코드의 복잡성을 줄이기 위한 도구라고 생각합니다.
3번 기능에 대해서, 출석, 지각, 결석, 경고 or 면담 or 제적 -> 4가지의 값을 전달해야하는 경우가 있습니다.
이러한 경우에 DTO를 사용하여 좀 더 관리하기 수월해지게 하는 용도라고 생각합니다. 
(또한, 전달해야하는 데이터 값이 필요없어 삭제되는 경우에도 이러한 비즈니스 로직이 바뀌지 안고, DTO만 바뀌게 된다는 장점이 있는 것 같아요.)

hyeonic replied:
dto 활용은 데이터 전달 목적으로 비즈니스 로직을 해치지 않고 dto만 수정하여 view에 요구하는 데이터를 가공할 수 있는 장점이 있을 것 같아요. 다만 추가적인 객체를 구성한다는 것 자체가 관리 포인트로 다가올 수도 있습니다. 제 의견은 이정도 규모의 요구사항이면 dto를 굳이 사용하지 않아도 괜찮다고 생각해요. 다만 중요한 것은 체체가 판단하였을 때 왜 더 복잡하지 않은 구조를 가지지 않는지를 명확히 설명할 수 있다면 적절히 활용하는 것도 좋습니다! 

말씀 드리고 싶은 부분을 정리하면 dto 사용 여부에 대해서 체체만의 적절한 근거를 가지고 있다면 사용하여도 괜찮다라고 이야기하고 싶어요.

hyeonic commented on src/main/java/attendance/dto/AttendanceContentDTO.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.List;
+import java.util.Set;
+
+public record AttendanceContentDTO(List<Attendance> attendances, Set<String> names) {
+}
```
record와 일반적인 class는 어떤 차이가 있을까요??

CheChe903 replied:
레코드에는 기능들이 내재 되어있다는 장점이 있습니다!
레코드를 사용한다면 보일러 플레이트 코드를 줄일 수 있다는 것이 저에게 사용되는 이유로 다가왔습니다.

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+import java.util.Locale;
+
+
+public class AttendanceController {
```
`AttendanceController`의 책임이 너무 막중한 것 같아요!

CheChe903 replied:
최대한 줄여보았지만,, 현재 고쳐진 코드 이후 좀 더 책임을 줄일 수 있는 방법이 있을까요 ?
 힌트로 주시면 감사하겠습니다!

hyeonic replied:
잘 수정해주셨네요 👍 다만 한 가지 우려되는 포인트가 있는 것 같아요. FUNCTION과 관련된 행위를 모으다 보니 `AttendanceFunctionExecutor`가 탄생한 것으로 보이는데요! 체체가 의도하지 않았지만 현재 두 객체의 패키지 의존성을 살펴보면 `상호 참조`하는 형태가 된 것 같아요

```java
import attendance.controller.AttendanceController; // 상호  참조
import attendance.domain.Function.AttendanceCheckFunction;
import attendance.domain.Function.AttendanceFunction;
import attendance.domain.Function.AttendanceHistoryFunction;
import attendance.domain.Function.AttendanceModifyFunction;
import attendance.domain.Function.QuitFunction;
import attendance.domain.Function.RiskOfExpulsionFunction;
import java.util.Map;

public class AttendanceFunctionExecutor { ... }
```

```java
import attendance.domain.Attendance;
import attendance.domain.AttendanceBook;
import attendance.domain.AttendanceFunctionExecutor; // 상호 참조
import attendance.domain.Time;
import attendance.dto.AttendanceContentDTO;
import attendance.repository.AttendanceRepository;
import attendance.utils.AttendanceReader;
import attendance.utils.FileReader;
import attendance.view.InputView;
import attendance.view.OutputView;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.TextStyle;
import java.util.List;
import java.util.Locale;
import java.util.stream.Stream;


public class AttendanceController { ... }
```

이렇게 된 이유는 각각의 Function이 수행하는 로직이 Controller 내의 여러 InputView, OutputView 등을 활용해야 하기 때문인 것으로 이해하였는데요, 객체가 `상호 참조`를 하였을 때 어떤 문제가 발생할 수 있을지 확인해보면 좋을 것 같아요! 현재 구조에서는 크게 복잡하지 않으니 AttendanceFunctionExecutor 및 AttendanceFunction 관련 책임을 Controller에게 위임하는 편도 좋아보이네요 ㅎㅎ

CheChe903 replied:
맞아요. 저도 수정하다가 기능 자체의 책임을 각각 기능 클래스에게 위임하고 싶었지만, 
그렇게 되면 인풋 또는 아웃풋 뷰, 레파지토리, 출석부 등등, 필요한 객체들을 각각 클래스에 주입하거나, 필드로 갖고 있어서 그게 마음에 들지 않았어요.
지금 좀 더 고민해보겠습니다!

hyeonic reviewed (changes requested):
안녕하세요 체체! 출석 리뷰를 맡은 매트입니다 😃 질문 주신 부분 코드 내에 남겨두었습니다!

제가 드린 질문이 모호할 수 있어서 이해가 안가는 부분이 있다면 언제든지 추가 코멘트 혹은 DM 부탁드립니다!

CheChe903 committed:
refactor(AttendanceController): indent 1로 측소

CheChe903 committed:
refactor: 기능 추상화 진행

if문으로 이루어지던 기능 선택을 추상화를 통해 분리합니다.
또한 start 메서드의 indent depth를 1줄입니다.

CheChe903 committed:
refactor: 메서드 순서 컨벤션

CheChe903 committed:
refactor: 파일 개행 추가

CheChe903 committed:
refactor: 매직넘버 상수화 진행

CheChe903 committed:
refactor(AttendanceController): stream으로 메서드 변경

CheChe903 committed:
refactor(AttendanceController): stream으로 메서드 변경, 파일경로 상수화

hyeonic commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,46 @@
+        return value;
+    }
+
+    public static String getAcademicStatus(int late, int absent) {
+        return Stream.of(late / STANDARD_OF_CHANGE_ABSENCE + absent)
+                .map(count -> {
+                    if (count > EXPELLED_COUNT) {
+                        return EXPELLED.getValue();
+                    }
+                    if (count >= INTERVIEW_COUNT) {
+                        return INTERVIEW.getValue();
+                    }
+                    if (count == WARNING_COUNT) {
+                        return WARNING.getValue();
+                    }
+                    return NOT.getValue();
+                })
+                .findFirst()
+                .orElse(NOT.getValue());
+    }
```
반복된 if 문을 줄여볼 순 없을까요?? 앞서 체체가 작성했던 functions의 개념을 활용해봐도 좋을 것 같아요. 힌트를 좀 더 드리면 Java에서는 기본적으로 상황에 맞게 사용할 수 있는 [Functional Interface](https://www.baeldung.com/java-8-functional-interfaces)가 존재하는데 이를 enum에서 필드로 의존하거나 하는 형태로요 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.stream.Stream;
+
+public enum AcademicStatus {
```
단위 테스트가 추가되면 좋을 것 같아요!

hyeonic commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,76 @@
+import java.time.format.TextStyle;
+import java.util.Locale;
+
+public record Time(LocalDate date, String hour, String minute, boolean isAbsent) {
+    public Time {
+        if (!isAbsent) {
+            validatePossibleTime(date, hour, minute);
+            validateInRangeTime(hour, minute);
+        }
+    }
```
`LocalDate` 대신 `LocalDateTime`을 활용할 순 없을까요??

CheChe903 replied:
제가 `LocalDate`를 사용한 이유가 있어요.
3번 기능은 크루의 닉네임으로 출석을 조회하는 기능입니다.
여기서 출석하지 않은 날은 시간 데이터가 없고 날짜만 존재하게 됩니다.
view에는 결석한 날을 `2025-02-24 --:-- (결석)` 이런 식으로 보여줘야 해서, 시간과 분을 `String`으로 받아서 처리했어요. 
근데 이 부분에 대해 다른 크루한테 이야기했더니, 도메인이 화면,`view`에 의존하는 것 같다는 피드백을 받았어요.
저는 `'--:--'`를 일종의 기본값으로 생각해서 이렇게 구현했는데요, 출석한 경우와 결석한 경우를 구분하기 위해 LocalDate를 사용했어요.
혹시 이런 구현 방식이 도메인이 `view`에 의존하는 걸까요? 매트의 의견도 궁금해요.

hyeonic replied:
이건 결석을 어떻게 처리하냐에 따라 달라질 것 같아요. 결석이라는 부분도 관리하고자 하는 데이터로 포함하게 되면 별도 결석을 나타내기 위한 별도 flag 필드를 두거나 체체가 작성하신 방법을 고려할 수 있을 것 같아요. 혹은 결석 자체를 관리하는 않는 것도 방법이겠죠? View의 의존 여부를 판단하기 이전에 먼저 결석을 도메인 내에서 어떻게 판단할지 고민해보는건 어떨까요? 그걸 View에 표현하는 것은 후순위가 아닐까 생각이 드네요.

hyeonic commented on src/main/java/attendance/repository/AttendanceRepository.java:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    public List<Attendance> findAllAttendanceByName(final String name) {
+        return attendances.stream()
+                .filter(attendance -> attendance.getCrewName().equals(name))
+                .filter(attendance -> attendance.getAttendanceTime().getMonth() == LocalDate.now().getMonthValue())
+                .sorted(Comparator.comparingInt(attendance -> attendance.getAttendanceTime().getDay()))
+                .toList();
+    }
```
findAllAttendanceByName를 테스트하려면 내부에 LocalDate.now()로 인해 무조건 현재 month를 기준으로 테스트가 진행되기 때문에 일관성 있는 테스트가 어려울 수 있을 것 같아요! 가령 현재 2025년 2월 기준으로 테스트 데이터를 구성하였지만 2026년 1월에 해당 테스트를 돌리면 성공을 보장할 수 있을까요?

hyeonic commented on src/main/java/attendance/utils/AttendanceReader.java:
```diff
@@ -0,0 +1,39 @@
+            String[] split = content.split(",");
+
+            String crewName = split[0];
+            LocalDateTime attendanceTime = LocalDateTime.parse(split[1],
+                    DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
```
`DateTimeFormatter`는 재사용이 가능하지 않을까요? 이번 기회에 `불변 객체`는 무엇이며 어떤 특성을 가지는지 알아봐도 좋을 것 같아요 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/Function/AttendanceCheckFunction.java:
```diff
@@ -0,0 +1,18 @@
+package attendance.domain.Function;
+
+import attendance.controller.AttendanceController;
+
+public class AttendanceCheckFunction implements AttendanceFunction {
+
+    private final AttendanceController attendanceController;
+
+    public AttendanceCheckFunction(AttendanceController attendanceController) {
+        this.attendanceController = attendanceController;
+    }
+
+    @Override
+    public boolean execute() {
+        attendanceController.attendanceCheckFunction();
+        return false;
+    }
+}
```
java 패키지 컨벤션에 맞춰서 `attendance.domain.Function`두 수정되면 좋을 것 같아요 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/Function/AttendanceFunction.java:
```diff
@@ -0,0 +1,6 @@
+package attendance.domain.Function;
+
+public interface AttendanceFunction {
+
+    boolean execute();
+}
```
`@FunctionalInterface` 애노테이션에 대해 알아보면 좋을 것 같아요! 추가로 boolean 타입을 반환하는 행위를 가진 Functional Interface도 존재하는데요, 직접 만들어 사용하는 것도 좋지만 java에서 기본 제공하는 Functional Interface를 알아보고 적절히 적용하는 것도 좋아보입니다 ㅎㅎ

CheChe903 replied:
매트 덕분에 뭔가 깨달음을 얻은 느낌이네요 ㅋㅋㅋ 감사합니다 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public boolean isAfter(LocalDateTime localDateTime) {
+        if (!isAbsent) {
+            return date.atTime(Parser.parseInt(hour), Parser.parseInt(minute)).isAfter(localDateTime);
+        }
+
+        return true;
+    }
```
인텔리제이에서 주고 있는 힌트 잘 해석해서 수정하면 좋을 것 같아요! 메서드가 의도와 다르게 동작하고 있는 것으로 보입니다. 테스트 코드가 추가되면 더욱 좋겠죠?!

![image](https://github.com/user-attachments/assets/4cfc2c33-3289-4d1b-a878-546633c895ae)

hyeonic commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,76 @@
+        validateAttendanceTime(hour, minute);
+    }
+
+    private static void validateAttendanceTime(String hour, String minute) {
+        int parsingHour = Parser.parseInt(hour);
+        int parsingMinute = Parser.parseInt(minute);
+
+        if (parsingHour < 8 || parsingHour == 23 && parsingMinute > 0) {
+            throw new IllegalArgumentException("[ERROR] 출석 가능한 시간이 아닙니다.");
+        }
+    }
+
+    private static void validateAttendanceDate(LocalDate date) {
+        String day = date.getDayOfWeek().name();
+
+        if (day.equals("SATURDAY") || day.equals("SUNDAY")) {
+            String message = String.format("[ERROR] %02d월 %02d일 %s은 등교일이 아닙니다.",
+                    date.getMonthValue(), date.getDayOfMonth(),
+                    date.getDayOfWeek().getDisplayName(TextStyle.FULL, Locale.KOREAN));
+            throw new IllegalArgumentException(message);
+        }
+    }
```
두 메서드는 static일 필요가 있을까요?

CheChe903 replied:
이동하면서 생긴 실수이네요! 꼼꼼히 봐주셔서 정말 감사합니다 ㅠㅠ

hyeonic commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,76 @@
+    private static void validateAttendanceDate(LocalDate date) {
+        String day = date.getDayOfWeek().name();
+
+        if (day.equals("SATURDAY") || day.equals("SUNDAY")) {
```
`DayOfWeek`는 enum인데요! name를 통해 equals 비교를 할 필요가 있을까요? 객체 자체를 비교할 순 없는 걸까요? enum의 ==나 equals의 특성에 대해 알아보면 좋을 것 같아요!

hyeonic reviewed (changes requested):
안녕하세요 체체! 답변 확인 하였고 추가 코멘트 남겼습니다 ㅎㅎ

모호하거나 이해안가는 부분 있다면 편하게 DM 부탁드려요!

CheChe903 committed:
refactor(AttendanceController): 순환 참조 삭제 리팩터링

순환 참조가 되던 문제를 해결합니다.
서플라이어를 이용하여 컨트롤러 자체에서 매개변수 없는 boolean 값을 반환하며 기능을 재발동합니다.

CheChe903 committed:
refactor(AttendanceStatus): 반복되는 if문 제거

Predicate를 사용하여 ENUM안에 컨디션을 넣습니다.

CheChe903 committed:
feat(AcademicStatus): 단위 테스트 진행

CheChe903 committed:
refactor: 각 기능 별 날짜 주입으로 해결

기존에는 각 기능들 자체에서 LocalDate.now()를 가져 테스트가 실패하는 경우가 생기는 가능성이 있었습니다.
따라서 날짜를 주입함으로써 테스트를 용이하게 합니다.

CheChe903 committed:
refactor(AttendanceReader): 데이터타임 포매터 상수화

CheChe903 committed:
refactor(Time): isAfter invert 경고 해결

CheChe903 committed:
refactor(Time): invert 오류 해결

CheChe903 committed:
refactor(Time): 검증 메서드 리팩터링

CheChe903 commented:
매트에게 궁금한 것이 있습니다 !

페어 프로그래밍을 하며 깜빡하고 있었지만, 현재 주말은 등교일이 아니라고 나오지만, 공휴일에 대해서는 처리하지 못했습니다.
이러한 경우에 대한 매트는 어떻게 처리할 지 힌트를 받아보고 싶습니다.

꼼꼼히 제 실수들을 봐주셔서 감사합니다 ~!

CheChe903 committed:
refactor(AttendanceStatus): String -> ENUM 반환으로 변경

기존에는 비즈니스 로직에도 ENUM 값이 아닌 ENUM VALUE를 반환했습니다.
이제는 ENUM 자체를 반환하도록 리팩터링합니다.
이유
1. 비즈니스 로직의 안전성
2. 값 캡슐화

hyeonic commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    private Map<String, Supplier<Boolean>> initializeFunctions() {
+        return Map.of(
+                CHECK_FUNCTION, () -> executeWithExceptionHandling(this::attendanceCheckFunction),
+                MODIFY_FUNCTION, () -> executeWithExceptionHandling(this::attendanceModifyFunction),
+                HISTORY_FUNCTION, () -> executeWithExceptionHandling(this::attendanceHistoryByNameFunction),
+                RISK_OF_EXPULSION_FUNCTION, () -> executeWithExceptionHandling(this::crewAtRiskOfExpulsion),
+                QUIT_FUNCTION, () -> true
+        );
+    }
+
+    private boolean executeWithExceptionHandling(Supplier<Boolean> function) {
+        try {
+            return function.get();
+        } catch (IllegalArgumentException e) {
+            outputView.printErrorMessage(e.getMessage());
+            return false;
+        }
+    }
```
훨씬 깔끔해진 것 같네요 ㅎㅎ 👍 함수라는 `행위 전달`을 통해서도 `전략 패턴`과 유사하게 개발이 가능하다는 것을 염두해두면 좋을 것 같아요. Supplier를 예로 들면 결국 수행하고자 하는 행위를 get으로 적절히 추상화하고 그에 대한 구현체를 상황 혹은 전략에 맞게 주입할 수 있으니까요!

추가로 `확정된 값`을 전달하는 것과 `함수를 통해 행위를 전달하는 것`에 대한 시점 차이도 잘 알아두면 좋을 것 같아요. 약간의 힌트를 드리면 `Lazy Evaluation`이란 키워드로 학습해보면 좋을 것 같아요!

hyeonic commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,55 @@
+        return AttendanceStatus.checkStatusWithCondition(attendanceTime, EDUCATION_HOUR, LATE_MINUTE, ABSENT_MINUTE);
+    }
+
+    public boolean isAlreadyAttendance(final Attendance currentAttendance) {
+
+        if (!crewName.equals(currentAttendance.crewName)) {
+            return false;
+        }
+
+        return attendanceTime.date().isEqual(currentAttendance.attendanceTime.date());
+    }
```
저는 보통 if 내의 조건은 긍정문이 오는 것을 선호합니다. 시간 날 때 `클린코드 17장 냄새와 휴리스틱`을 읽어보면 도움이 될 것 같아요!

```java
    public boolean isAlreadyAttendance(final Attendance currentAttendance) {
        if (crewName.equals(currentAttendance.crewName)) {
            return attendanceTime.date().isEqual(currentAttendance.attendanceTime.date());
        }

        return false;
    }

```

hyeonic commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public List<CrewNameAndAcademicStatusDTO> getCrewAtRiskOfExpulsion(
+            final AttendanceRepository attendanceRepository,
+            final String academicStatus,
+            final int month) {
+
+        return names.stream()
+                .map(name -> attendanceRepository.getAcademicStatusByName(name, month))
+                .filter(dto -> dto.academicStatus().equals(academicStatus))
+                .collect(Collectors.toList());
+    }
```
체체가 정의한 DTO는 도메인에 가까울까요, View에 가까울까요? 만약 출력 형식이 변경되면 DTO는 변경에서 자유로울 수 있을까요? 도메인 내에 DTO 변환 로직은 그러한 변경을 피할 수 있을까요?

hyeonic commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,36 @@
+
+public enum AcademicStatus {
+
+    EXPELLED("제적", count -> count > 5),
+    INTERVIEW("면담", count -> count >= 3),
+    WARNING("경고", count -> count == 2),
+    NOT("없음", count -> count < 2);
```
적절하게 책임을 잘 위치 시킨 것 같아요 😃

hyeonic commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,36 @@
+        return Stream.of(values())
+                .filter(status -> status.determineStatusConditions.test(totalCount))
+                .findFirst()
+                .orElse(NOT)
```
`orElse`와 `orElseGet`은 어떠한 차이가 있을까요? 판단된 값과 함수의 행위에 대해 고민해보면 좋을 것 같아요!

hyeonic commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,36 @@
+                .filter(status -> status.determineStatusConditions.test(totalCount))
+                .findFirst()
+                .orElse(NOT)
+                .getValue();
```
어쩌면 도메인 내에 View에 필요한 출력 데이터를 가지고 있는 것 처럼 보이기도 하네요!

hyeonic commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+
+    public boolean isSameByNameAndLocalDate(final String name, int year, int month, int day) {
+        return crewName.equals(name) && day == attendanceTime.getDay() && year == attendanceTime.getYear()
+                && month == attendanceTime.getMonth();
+    }
```
파라미터에 `final`을 넣고 싶다면 일관성 있게 추가하는 편은 어떨까요??

hyeonic commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,40 @@
+    public static AttendanceStatus checkStatusWithCondition(Time attendanceTime, int hour, int lateMinute,
+                                                            int absentMinute) {
+
+        if (attendanceTime.isBefore(
+                LocalDateTime.of(attendanceTime.getYear(), attendanceTime.getMonth(), attendanceTime.getDay(),
+                        hour,
+                        lateMinute))) {
+            return ATTEND;
+        }
+
+        if (attendanceTime.isBefore(
+                LocalDateTime.of(attendanceTime.getYear(), attendanceTime.getMonth(), attendanceTime.getDay(),
+                        hour,
+                        absentMinute))) {
+            return LATE;
+        }
```
이것도 반복된 if 문으로 볼 수 있을 것 같아요! `AcademicStatus`와 유사하게 수정할 수도 있겠네요 ㅎㅎ

hyeonic commented on src/main/java/attendance/repository/AttendanceRepository.java:
```diff
@@ -0,0 +1,95 @@
+public class AttendanceRepository {
+    private final List<Attendance> attendances;
+
+    public AttendanceRepository(final List<Attendance> attendances) {
+        this.attendances = attendances;
+    }
```
만약 외부에서 주입된 List에 엘리먼트가 추가되면 어떻게 될까요? [방어적 복사](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/)에 대해 알아보면 좋을 것 같아요!

hyeonic commented on src/main/java/attendance/repository/AttendanceRepository.java:
```diff
@@ -0,0 +1,95 @@
+                .forEach(date -> attendances.add(new Attendance(name, new Time(date, "--", "--", true))));
+    }
+
+    private boolean isWeekend(final LocalDate date) {
+        DayOfWeek dayOfWeek = date.getDayOfWeek();
+        return dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
+    }
```
> 현재 주말은 등교일이 아니라고 나오지만, 공휴일에 대해서는 처리하지 못했습니다.
이러한 경우에 대한 매트는 어떻게 처리할 지 힌트를 받아보고 싶습니다.

우리나라의 공휴일을 제공하는 API나 외부라이브러리를 통해 이를 표현하고 싶지만... 그게 쉽지 않다면 양이 그리 많지 않으니 상수를 통해 관리해야하지 않을까 생각이 드네요! 요부분은 구현하게 어렵다 느끼시면 자유롭게 고민해보고 판단해서 적용해보세요. 다만 너무 많은 시간을 쏟게 만든다면 주말만 잘 판단하고 다른 부분을 보완하는 편이 더 좋아보여요!

hyeonic commented on src/main/java/attendance/utils/AttendanceReader.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+import java.util.Set;
+
+public final class AttendanceReader {
+
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+
+    private AttendanceReader() {
+    }
+
+
+    public static AttendanceContentDTO getAttendanceRecordContent(final List<String> attendanceContents) {
+        attendanceContents.removeFirst();
+        final List<Attendance> attendances = new ArrayList<>();
+        final Set<String> names = new HashSet<>();
+
+        for (String content : attendanceContents) {
+            String[] split = content.split(",");
+
+            String crewName = split[0];
+            LocalDateTime attendanceTime = LocalDateTime.parse(split[1], formatter);
+
+            LocalDate localDate = attendanceTime.toLocalDate();
+            String hour = String.format("%02d", attendanceTime.getHour());
+            String minute = String.format("%02d", attendanceTime.getMinute());
+
+            names.add(crewName);
+            attendances.add(new Attendance(crewName, new Time(localDate, hour, minute, false)));
+        }
+        return new AttendanceContentDTO(attendances, names);
+    }
+}
```
싱글턴 패턴을 구현하셨네요! 어떠한 장단점이 있을까요??

CheChe903 replied:
페어가 알려준 방법이긴 한데, 이것도 싱글톤 패턴이라고 부를 수 있는 건지 의문이 듭니다!
페어가 말해주길, 유틸 클래스는 접근을 바로 하기 때문에 생성자가 필요없어 private으로 의미 없는 생성을 막는다라고 알려주었습니다!
매트의 생각이 궁금합니다!

hyeonic reviewed (approved):
안녕하세요 체체! 처음 리뷰 요청했을 때 보다 많이 깔끔해진 것 같아요 ㅎㅎ 고생하셨습니다 😃

2단계는 1단계 + 리뷰를 기반으로 TDD를 적극활용하여 `출석 다시 구현하기`를 하는 것으로 알고 있는데요! 출석 도메인에 대한 이해도가 많이 올라간 것 같아서 이번에는 적극적으로 TDD를 도입하면서 부족했던 단위 테스트도 잘 보완되면 좋을 것 같아요. 

간단히 코멘트 남겼습니다! 1단계는 이만 마무리하고 2단계에서 추가 고려해주세요 ㅎㅎ 다음 단계도 잘 부탁드립니다 😃
